lib/libutee/tee_api_operations.c

```diff
@@ -1224,46 +1273,69 @@ TEE_Result TEE_AEDecryptFinal(TEE_OperationHandle op,
 	uint32_t tmp_dlen;
 	size_t req_dlen;
 
-	if (op == TEE_HANDLE_NULL || (srcData == NULL && srcLen != 0) ||
-	    destLen == NULL || (destData == NULL && *destLen != 0) ||
-	    (tag == NULL && tagLen != 0))
-		TEE_Panic(0);
-	if (op->info.operationClass != TEE_OPERATION_AE)
-		TEE_Panic(0);
-	if ((op->info.handleState & TEE_HANDLE_FLAG_INITIALIZED) == 0)
-		TEE_Panic(0);
+	if (operation == TEE_HANDLE_NULL ||
+	    (srcData == NULL && srcLen != 0) ||
+	    destLen == NULL ||
+	    (destData == NULL && *destLen != 0) ||
+	    (tag == NULL && tagLen != 0)) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if (operation->info.operationClass != TEE_OPERATION_AE) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if ((operation->info.handleState & TEE_HANDLE_FLAG_INITIALIZED) == 0) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
 	/*
 	 * Check that required destLen is big enough before starting to feed
 	 * data to the algorithm. Errors during feeding of data are fatal as we
 	 * can't restore sync with this API.
 	 */
-	req_dlen = op->buffer_offs + srcLen;
+	req_dlen = operation->buffer_offs + srcLen;
 	if (*destLen < req_dlen) {
 		*destLen = req_dlen;
-		return TEE_ERROR_SHORT_BUFFER;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto out;
 	}
 
 	tmp_dlen = *destLen - acc_dlen;
-	tee_buffer_update(op, utee_authenc_update_payload, srcData, srcLen,
-			  dst, &tmp_dlen);
+	res = tee_buffer_update(operation, utee_authenc_update_payload, srcData,
+				srcLen, dst, &tmp_dlen);
+	if (res != TEE_SUCCESS)
+		goto out;
+
 	dst += tmp_dlen;
 	acc_dlen += tmp_dlen;
 
 	tmp_dlen = *destLen - acc_dlen;
-	res =
-	    utee_authenc_dec_final(op->state, op->buffer, op->buffer_offs, dst,
-				   &tmp_dlen, tag, tagLen);
-	if (res != TEE_SUCCESS && res != TEE_ERROR_MAC_INVALID)
-		TEE_Panic(res);
+	res = utee_authenc_dec_final(operation->state, operation->buffer,
+				     operation->buffer_offs, dst, &tmp_dlen,
+				     tag, tagLen);
+	if (res != TEE_SUCCESS)
+		goto out;
+
 	/* Supplied tagLen should match what we initiated with */
-	if (tagLen != op->ae_tag_len)
+	if (tagLen != operation->ae_tag_len) {
 		res = TEE_ERROR_MAC_INVALID;
+		goto out;
```

Comment:
@jenswi-linaro :  The spec is not really clear about this. Why do you think it is better to consider the operation finalized in that case (same as previous impletementation) since we have an error message (compared to other error messages such as TEE_ERROR_SHORT_BUFFER) ?